agent/agenttest/agent.go

```diff
@@ -0,0 +1,97 @@
+package agenttest
+
+import (
+	"context"
+	"net/url"
+	"sync"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/agent"
+	"github.com/coder/coder/v2/coderd/coderdtest"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/agentsdk"
+)
+
+// Agent is a small wrapper around an agent for use in tests.
+type Agent struct {
+	agent       agent.Agent
+	agentClient *agentsdk.Client
+	resources   []codersdk.WorkspaceResource
+	waiter      func(*codersdk.Client, uuid.UUID, ...string) []codersdk.WorkspaceResource
+	waitOnce    sync.Once
+}
+
+// Wait waits for the agent to connect to the workspace and returns the
+// resources for the connected workspace.
+// Calls coderdtest.AwaitWorkspaceAgents under the hood.
+// Multiple calls to Wait() are idempotent.
+func (a *Agent) Wait(client *codersdk.Client, workspaceID uuid.UUID, agentNames ...string) []codersdk.WorkspaceResource {
+	a.waitOnce.Do(func() {
+		a.resources = a.waiter(client, workspaceID, agentNames...)
```

Comment:
it smells funny to me that you pass in arguments that are accepted the first time this function is called, but ignored on subsequent calls.

Furthermore, what's the advantage of this method over just directly calling `AwaitWorkspaceAgents()`?

Reply:
> it smells funny to me that you pass in arguments that are accepted the first time this function is called, but ignored on subsequent calls.

👍 I second that, this could be error-prone, and easy to miss.

BTW usually in those cases, I used to name the function `...Once` to highlight the presence of `sync.Once`.

Reply:
The overall consensus here seems to be to keep using `coderdtest.AwaitWorkspaceAgents`